### PR TITLE
feat: introduce StateWriter and HashedStateWriter traits

### LINKED ISSUE #6570: Introduce Writer provider traits for BundleStateWithReceipts and related types
Similar to https://github.com/paradigmxyz/reth/pull/3384, all existing structs that wrap other structs and have a single `write_to_db` method should be replaced by a provider trait. These structs are:
* `HashedStateChanges`
* `BundleStateWithReceipts`
* `StateReverts`
* `StateChanges`

ref https://github.com/paradigmxyz/reth/issues/6461#issuecomment-1939513765

### PR DESCRIPTION
**Pull request by Rjected**
Fixes https://github.com/paradigmxyz/reth/issues/6570

Removes the `StateChanges`, `StateReverts`, and `HashedStateChanges` types. Replaces their `write_to_db` methods with methods on two new provider traits.

These traits are `StateWriter` and `HashedStateWriter`. Introduces an `update` method which functions like the `Database::update` method, but on the `DatabaseProviderRW`. This is mainly for `reth drop-stage`.

### CODE REVIEW

File: crates/storage/provider/src/bundle_state/state_reverts.rs [hunk 1]
```diff
@@ -1,108 +1,33 @@
...
+where
+    R: Iterator<Item = A>,
+    W: Iterator<Item = B>,
+    A: std::fmt::Debug,
+    B: std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("StorageRevertsIter").finish()
+    }
+}
```
Comment (mattsse) [inline crates/storage/provider/src/bundle_state/state_reverts.rs:22]: isn't this the same as deriving debug but now without including the conent

File: crates/storage/provider/src/providers/database/provider.rs [hunk 2]
```diff
@@ -67,6 +72,40 @@ pub type DatabaseProviderRO<DB> = DatabaseProvider<<DB as Database>::TX>;
...
+    pub fn with_tx(
+        tx: <DB as Database>::TXMut,
+        chain_spec: Arc<ChainSpec>,
+        snapshot_provider: &Option<Arc<SnapshotProvider>>,
+    ) -> ProviderResult<DatabaseProviderRW<DB>> {
+        let mut provider = DatabaseProvider::new_rw(tx, chain_spec.clone());
+
+        if let Some(snapshot_provider) = snapshot_provider {
+            provider = provider.with_snapshot_provider(snapshot_provider.clone());
+        }
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:91]: there should also be a function on the db provider that accepts all args

File: crates/storage/provider/src/providers/database/provider.rs [hunk 3]
```diff
@@ -67,6 +72,40 @@ pub type DatabaseProviderRO<DB> = DatabaseProvider<<DB as Database>::TX>;
...
+
+        Ok(DatabaseProviderRW(provider))
+    }
+
+    /// Takes a function and passes a read-write transaction into it, making sure it's committed at
+    /// the end of the execution.
+    pub fn update<T, F>(mut self, f: F) -> Result<T, DatabaseError>
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:98]: I think we also want to name this commit_ ?

**Review by mattsse — CHANGES_REQUESTED** [review]
I think these traits make sense

wdyt @rkrasiuk

File: crates/storage/provider/src/bundle_state/state_reverts.rs (hunk 1 above)
Comment (Rjected) [reply crates/storage/provider/src/bundle_state/state_reverts.rs:22]: yeah, can't derive debug though, mainly did this for the missing debug impl lint on public items. Does debug make sense here? I can add contents if so

File: crates/storage/provider/src/providers/database/provider.rs (hunk 2 above)
Comment (Rjected) [reply crates/storage/provider/src/providers/database/provider.rs]: just added one, this method now calls that method

**Review by mattsse — APPROVED** [review]
lgtm pending @rkrasiuk

**Review by rkrasiuk — COMMENTED** [review]
left some nits, comments & request for follow up, neither is a blocker. i didn't review write fns closely since they are only moved from previous loc, no refactor/logic change was done. lgtm!

**Review by rkrasiuk — APPROVED** [review]

**Comment by shekhirin** [comment]
@Rjected do we still want this?

**Comment by Rjected** [comment]
> @Rjected do we still want this?

yes, I need to fix merge conflicts though... will finish soon

**Pull request closed without merging**